From a7c7cfffdbbb0f2ecb4e0e7fd82ca6556d05209e Mon Sep 17 00:00:00 2001
From: Benjamin Herrenschmidt <benh@...>
Date: Mon, 1 Jun 2026 14:05:09 -0400
Subject: [PATCH] Add support for openssl 3 providers and building without
 engines

This makes things build when engine support is completely disabled
in openssl (ie, Fedora currently applies a downstream patch) and
adds support for the "store" API when no engine is selected which
works with providers.

Signed-off-by: Benjamin Herrenschmidt <benh@...>
Signed-off-by: James Bottomley <James.Bottomley@...>

Upstream-Status: Submitted [https://groups.io/g/sbsigntools/topic/patch_1_4_add_support_for/119597824]
Signed-off-by: Ross Burton <ross.burton@arm.com>
---
 src/fileio.c    | 91 ++++++++++++++++++++++++++++++++++++++++---------
 src/fileio.h    |  5 +++
 src/sbsign.c    | 30 +++++++++++++++-
 src/sbvarsign.c | 30 +++++++++++++++-
 4 files changed, 137 insertions(+), 19 deletions(-)

diff --git a/src/fileio.c b/src/fileio.c
index 929bb00..1957ab0 100644
--- a/src/fileio.c
+++ b/src/fileio.c
@@ -39,7 +39,13 @@
 #include <openssl/bio.h>
 #include <openssl/pem.h>
 #include <openssl/err.h>
+#include <openssl/ui.h>
+#ifndef OPENSSL_NO_ENGINE
 #include <openssl/engine.h>
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+#include <openssl/store.h>
+#endif
 
 #include <ccan/talloc/talloc.h>
 #include <ccan/read_write_all/read_write_all.h>
@@ -60,29 +66,39 @@ static int ui_read(UI *ui, UI_STRING *uis)
 	return 1;
 }
 
-EVP_PKEY *fileio_read_engine_key(const char *engine, const char *filename)
+static UI_METHOD *ui_create(void)
 {
 	UI_METHOD *ui;
-	ENGINE *e;
-	EVP_PKEY *pkey = NULL;
-
-	ENGINE_load_builtin_engines();
-	e = ENGINE_by_id(engine);
-
-	if (!e) {
-		fprintf(stderr, "Failed to load engine: %s\n", engine);
-		ERR_print_errors_fp(stderr);
-		return NULL;
-	}
 
 	ui = UI_create_method("sbsigntools");
 	if (!ui) {
 		fprintf(stderr, "Failed to create UI method\n");
 		ERR_print_errors_fp(stderr);
-		goto out_free;
+		return NULL;
 	}
 	UI_method_set_reader(ui, ui_read);
 
+	return ui;
+}
+
+#ifndef OPENSSL_NO_ENGINE
+EVP_PKEY *fileio_read_engine_key(const char *engine, const char *filename)
+{
+	EVP_PKEY *pkey = NULL;
+	ENGINE *e = NULL;
+	UI_METHOD *ui;
+
+	ENGINE_load_builtin_engines();
+	ui = ui_create();
+	if (!ui)
+		return NULL;
+	e = ENGINE_by_id(engine);
+	if (!e) {
+		fprintf(stderr, "Failed to load engine: %s\n", engine);
+		ERR_print_errors_fp(stderr);
+		goto out_free;
+	}
+
 	if (!ENGINE_init(e)) {
 		fprintf(stderr, "Failed to initialize engine %s\n", engine);
 		ERR_print_errors_fp(stderr);
@@ -90,13 +106,55 @@ EVP_PKEY *fileio_read_engine_key(const char *engine, const char *filename)
 	}
 
 	pkey = ENGINE_load_private_key(e, filename, ui, NULL);
+	if (!pkey) {
+		fprintf(stderr, "Can't load key '%s' from engine '%s'\n",
+			filename, engine);
+		ERR_print_errors_fp(stderr);
+	}
 	ENGINE_finish(e);
 
  out_free:
 	ENGINE_free(e);
+	UI_destroy_method(ui);
 	return pkey;
 }
+#endif
 
+#if OPENSSL_VERSION_MAJOR >= 3
+EVP_PKEY *fileio_read_pkey(const char *filename)
+{
+	EVP_PKEY *key = NULL;
+	OSSL_STORE_CTX *store_ctx = NULL;
+	OSSL_STORE_INFO *info = NULL;
+	UI_METHOD *ui;
+
+	ui = ui_create();
+	if (!ui)
+		return NULL;
+
+	store_ctx = OSSL_STORE_open(filename, ui, NULL, NULL, NULL);
+	if (!store_ctx) {
+		fprintf(stderr, "error opening store %s\n", filename);
+		goto out_free;
+	}
+	while (!key && !OSSL_STORE_eof(store_ctx)) {
+		info = OSSL_STORE_load(store_ctx);
+		if (!info)
+			continue;
+		if (OSSL_STORE_INFO_get_type(info) == OSSL_STORE_INFO_PKEY)
+			key = OSSL_STORE_INFO_get1_PKEY(info);
+		OSSL_STORE_INFO_free(info);
+	}
+	if (!key) {
+		fprintf(stderr, "Can't load key from url '%s'\n", filename);
+		ERR_print_errors_fp(stderr);
+	}
+out_free:
+	OSSL_STORE_close(store_ctx);
+	UI_destroy_method(ui);
+	return key;
+}
+#else
 EVP_PKEY *fileio_read_pkey(const char *filename)
 {
 	EVP_PKEY *key = NULL;
@@ -104,18 +162,17 @@ EVP_PKEY *fileio_read_pkey(const char *filename)
 
 	bio = BIO_new_file(filename, "r");
 	if (!bio)
-		goto out;
-
+		return NULL;
 	key = PEM_read_bio_PrivateKey(bio, NULL, NULL, NULL);
-
-out:
 	BIO_free_all(bio);
 	if (!key) {
 		fprintf(stderr, "Can't load key from file '%s'\n", filename);
 		ERR_print_errors_fp(stderr);
 	}
+
 	return key;
 }
+#endif
 
 X509 *fileio_read_cert(const char *filename)
 {
diff --git a/src/fileio.h b/src/fileio.h
index b3ed22c..d7f2214 100644
--- a/src/fileio.h
+++ b/src/fileio.h
@@ -38,7 +38,12 @@
 #include <openssl/x509.h>
 
 EVP_PKEY *fileio_read_pkey(const char *filename);
+#ifndef OPENSSL_NO_ENGINE
 EVP_PKEY *fileio_read_engine_key(const char *engine, const char *filename);
+#else
+static inline EVP_PKEY *fileio_read_engine_key(const char *engine, const char *filename)
+{ (void)engine; (void)filename; return NULL; }
+#endif
 X509 *fileio_read_cert(const char *filename);
 
 int fileio_read_file(void *ctx, const char *filename,
diff --git a/src/sbsign.c b/src/sbsign.c
index 3a5ed32..c32d9c5 100644
--- a/src/sbsign.c
+++ b/src/sbsign.c
@@ -52,6 +52,10 @@
 #include <openssl/bio.h>
 #include <openssl/x509.h>
 
+#if OPENSSL_VERSION_MAJOR >= 3
+#include <openssl/provider.h>
+#endif
+
 #include <ccan/talloc/talloc.h>
 
 #include "idc.h"
@@ -77,7 +81,12 @@ static struct option options[] = {
 	{ "verbose", no_argument, NULL, 'v' },
 	{ "help", no_argument, NULL, 'h' },
 	{ "version", no_argument, NULL, 'V' },
+#ifndef OPENSSL_NO_ENGINE
 	{ "engine", required_argument, NULL, 'e'},
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+	{ "provider", required_argument, NULL, 'p'},
+#endif
 	{ "addcert", required_argument, NULL, 'a'},
 	{ NULL, 0, NULL, 0 },
 };
@@ -88,7 +97,12 @@ static void usage(void)
 			"<efi-boot-image>\n"
 		"Sign an EFI boot image for use with secure boot.\n\n"
 		"Options:\n"
+#ifndef OPENSSL_NO_ENGINE
 		"\t--engine <eng>     use the specified engine to load the key\n"
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+		"\t--provider <prov>  load the specified provider\n"
+#endif
 		"\t--key <keyfile>    signing key (PEM-encoded RSA "
 						"private key)\n"
 		"\t--cert <certfile>  certificate (x509 certificate)\n"
@@ -169,7 +183,14 @@ int main(int argc, char **argv)
 
 	for (;;) {
 		int idx;
-		c = getopt_long(argc, argv, "xo:c:k:dvVhe:a:", options, &idx);
+		c = getopt_long(argc, argv, "xo:c:k:dvVha:"
+#ifndef OPENSSL_NO_ENGINE
+				"e:"
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+				"p:"
+#endif
+				, options, &idx);
 		if (c == -1)
 			break;
 
@@ -198,9 +219,16 @@ int main(int argc, char **argv)
 		case 'h':
 			usage();
 			return EXIT_SUCCESS;
+#ifndef OPENSSL_NO_ENGINE
 		case 'e':
 			engine = optarg;
 			break;
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+		case 'p':
+			OSSL_PROVIDER_load(NULL, optarg);
+			break;
+#endif
 		case 'a':
 			addcertfilename = optarg;
 			break;
diff --git a/src/sbvarsign.c b/src/sbvarsign.c
index 58031ec..adaf701 100644
--- a/src/sbvarsign.c
+++ b/src/sbvarsign.c
@@ -51,6 +51,10 @@
 #include <openssl/pem.h>
 #include <openssl/pkcs7.h>
 
+#if OPENSSL_VERSION_MAJOR >= 3
+#include <openssl/provider.h>
+#endif
+
 #include <ccan/array_size/array_size.h>
 #include <ccan/read_write_all/read_write_all.h>
 #include <ccan/talloc/talloc.h>
@@ -397,7 +401,12 @@ static struct option options[] = {
 	{ "verbose", no_argument, NULL, 'v' },
 	{ "help", no_argument, NULL, 'h' },
 	{ "version", no_argument, NULL, 'V' },
+#ifndef OPENSSL_NO_ENGINE
 	{ "engine", required_argument, NULL, 'e'},
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+	{ "provider", required_argument, NULL, 'p'},
+#endif
 	{ NULL, 0, NULL, 0 },
 };
 
@@ -409,7 +418,12 @@ void usage(void)
 			"<var-name> <var-data-file>\n"
 		"Sign a blob of data for use in SetVariable().\n\n"
 		"Options:\n"
+#ifndef OPENSSL_NO_ENGINE
 		"\t--engine <eng>     use the specified engine to load the key\n"
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+		"\t--provider <prov>  load the specified provider\n"
+#endif
 		"\t--key <keyfile>    signing key (PEM-encoded RSA "
 						"private key)\n"
 		"\t--cert <certfile>  certificate (x509 certificate)\n"
@@ -455,7 +469,14 @@ int main(int argc, char **argv)
 
 	for (;;) {
 		int idx;
-		c = getopt_long(argc, argv, "o:g:a:k:c:ivVhe:", options, &idx);
+		c = getopt_long(argc, argv, "o:g:a:k:c:ivVh"
+#ifndef OPENSSL_NO_ENGINE
+				"e:"
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+				"p:"
+#endif
+				, options, &idx);
 		if (c == -1)
 			break;
 
@@ -487,9 +508,16 @@ int main(int argc, char **argv)
 		case 'h':
 			usage();
 			return EXIT_SUCCESS;
+#ifndef OPENSSL_NO_ENGINE
 		case 'e':
 			engine = optarg;
 			break;
+#endif
+#if OPENSSL_VERSION_MAJOR >= 3
+		case 'p':
+			OSSL_PROVIDER_load(NULL, optarg);
+			break;
+#endif
 		}
 	}
 
-- 
2.43.0

